File: app/src/main/java/com/example/myappcontacts/presentation/contactslist/presenter/ContactsListPresenter.java
package com.example.myappcontacts.presentation.contactslist.presenter;

import android.util.Log;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.example.myappcontacts.App;
import com.example.myappcontacts.business.contactslist.IContactsListInteractor;
import com.example.myappcontacts.data.dao.contacts.db.ContactsModel;
import com.example.myappcontacts.di.contactslist.ContactsListModule;
import com.example.myappcontacts.presentation.contactslist.view.IContactsListView;

import java.util.List;
import java.util.UUID;

import javax.inject.Inject;

import io.reactivex.disposables.CompositeDisposable;

@InjectViewState
public class ContactsListPresenter extends MvpPresenter<IContactsListView> implements IContactsListPresenter {
    private static final String TAG = "MY_TAG";

    private CompositeDisposable mDisposer = new CompositeDisposable();

    @Inject
    IContactsListInteractor mContactsListInteractor;

    public ContactsListPresenter() {
        App.get().plusContactsListModule(new ContactsListModule()).inject(this);
    }


    @Override
    public void addContact() {
        mDisposer.add(mContactsListInteractor.addContact()
                .subscribe(this::onAddSuccess, this::onError));
    }

    private void onAddSuccess(UUID contactId) {
        getViewState().openContact(contactId);
    }

    private void onError(Throwable throwable) {
        Log.e(TAG, getClass().getSimpleName() + " onError " + throwable);
    }

    @Override
    public void deleteContact(UUID contactId) {
        mDisposer.add(mContactsListInteractor.deleteContact(contactId)
                .subscribe(this::onDeleteSuccess, this::onError));
    }

    private void onDeleteSuccess() {
        Log.i(TAG, getClass().getSimpleName() + " contact deleted");
    }

    @Override
    public void loadContactsList() {
        mDisposer.add(mContactsListInteractor.loadContactsList()
                .subscribe(this::onLoadSuccess, this::onError));
    }

    private void onLoadSuccess(List<ContactsModel> contactsList) {
        Log.i("MY_TAG", getClass().getSimpleName() + " onLoadSuccess сработал " + contactsList.size());
        getViewState().updateContactsList(contactsList);
    }

    @Override
    public void onItemSwiped(UUID contactId) {
        getViewState().deleteContact(contactId);
    }

    @Override
    public void onContactItemClicked(UUID contactId) {
        getViewState().openContact(contactId);
    }

    @Override
    public void onShowContactsMapClicked() {
        getViewState().openMap();
    }

    @Override
    public void onDestroy() {
        mDisposer.dispose();
        App.get().clearContactsListComponent();
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/myappcontacts/presentation/contactslist/presenter/ContactsListPresenter.kt
package com.example.myappcontacts.presentation.contactslist.presenter

import android.util.Log
import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.example.myappcontacts.App
import com.example.myappcontacts.business.contactslist.IContactsListInteractor
import com.example.myappcontacts.data.dao.contacts.db.ContactsModel
import com.example.myappcontacts.di.contactslist.ContactsListModule
import com.example.myappcontacts.presentation.contactslist.view.IContactsListView
import io.reactivex.disposables.CompositeDisposable
import java.util.*
import javax.inject.Inject

internal const val TAG = "TEST"

@InjectViewState
class ContactsListPresenter : MvpPresenter<IContactsListView>(), IContactsListPresenter {

    @Inject
    lateinit var contactsListInteractor: IContactsListInteractor

    private val disposer = CompositeDisposable()

    init {
        App.get().plusContactsListModule(ContactsListModule()).inject(this)
    }

    override fun addContact() {
        disposer.add(contactsListInteractor
                .addContact()
                .subscribe({ onAddSuccess(it) }, { onError(it) }))
    }

    private fun onAddSuccess(contactId: UUID) = viewState.openContact(contactId)

    private fun onError(throwable: Throwable) =
            Log.e(TAG, this::class.simpleName + " onError " + throwable)

    override fun deleteContact(contactId: UUID) {
        disposer.add(contactsListInteractor
                .deleteContact(contactId)
                .subscribe({ onDeleteSuccess() }, { onError(it) }))
    }

    private fun onDeleteSuccess() =
            Log.i(TAG, this::class.simpleName + " contact deleted")

    override fun loadContactsList() {
        disposer.add(contactsListInteractor
                .loadContactsList()
                .subscribe({ onLoadSuccess(it) }, { onError(it) }))
    }

    private fun onLoadSuccess(contactsList: List<ContactsModel>) {
        viewState.updateContactsList(contactsList)
        Log.i(TAG, "${this::class.simpleName}: onLoadSuccess сработал, загружено ${contactsList.size} контактов")
    }

    override fun onItemSwiped(contactId: UUID) {
        viewState.deleteContact(contactId)
    }

    override fun onContactItemClicked(contactsId: UUID) {
        viewState.openContact(contactsId)
    }

    override fun onShowContactsMapClicked() {
        viewState.openMap()
    }

    override fun onDestroy() {
        disposer.dispose()
        App.get().clearContactsListComponent()
        super.onDestroy()
    }
}